matrix-org/matrix-react-sdk
# Add jump to date functionality to date headers in timeline v2

v1 in https://github.com/matrix-org/matrix-react-sdk/pull/7317

---



Add jump to date functionality to date headers in timeline

Relies on this `matrix-js-sdk` PR to add MSC3030 `/timestamp_to_event` support: https://github.com/matrix-org/matrix-js-sdk/pull/2072

Fix https://github.com/vector-im/element-web/issues/7677

Part of MSC3030: https://github.com/matrix-org/matrix-doc/pull/3030

Experimental Synapse implementation added in https://github.com/matrix-org/synapse/pull/9445

<img width="672" alt="" src="https://user-images.githubusercontent.com/558581/150060664-79627573-f4fd-497c-b726-dc3485854bd0.png">



### Testing strategy

 1. Update your Synapse `homeserver.yaml` to enable the `msc3030_enabled` experimental feature
	```
	experimental_features:
	  msc3030_enabled: true
	```
 1. Refresh Element so the Element Labs feature shows up
 1. Enable the `Jump to date` Labs feature
 1. Check out the timeline for the jump to date headers


### Dev notes


Screen recordings:

 - Quicktime with
 - https://github.com/keycastr/keycastr

---

Combining React callback refs and ref-refs,

 - https://github.com/facebook/react/issues/13029
 - https://dev.to/thekashey/merging-refs-41l8
 - https://dev.to/thekashey/the-same-useref-but-it-will-callback-8bo
 - https://itnext.io/reusing-the-ref-from-forwardref-with-react-hooks-4ce9df693dd
 - https://dev.to/wojciechmatuszewski/mutable-and-immutable-useref-semantics-with-react-typescript-30c9

```js
const attachRef = (input) => {
    this.input = input;

    if (typeof this.props.inputRef === 'function') {
        this.props.inputRef(input);
    } else if(this.props.inputRef) {
        this.props.inputRef.current = input;
    }
}
```



### TODO

 - [x] Refactor rest of `MenuItems`, https://github.com/matrix-org/matrix-react-sdk/pull/7339#discussion_r767153040
 - [x] Clean up PR
 - [x] Split out `ContextMenu` `RovingTabIndex` refactor to another PR, https://github.com/matrix-org/matrix-react-sdk/pull/7353
 - [x] Separate PR: `/jumptodate 2021-12-10`, https://github.com/matrix-org/matrix-react-sdk/pull/7372
 - [ ] Separate PR: Sticky date headers so one is always visible to reference and available to use to jump, https://github.com/vector-im/element-web/issues/7575














---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.




































































Preview: https://61f2350539af3b54bfe30e9a--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.
